crates/stc_ts_types/src/type_id/destructure.rs

```diff
@@ -0,0 +1,37 @@
+use std::sync::atomic::{AtomicU32, Ordering::SeqCst};
+
+use stc_visit::Visit;
+use swc_common::{EqIgnoreSpan, TypeEq};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Visit)]
+pub struct DestructureId(u32);
+
+impl DestructureId {
+    pub fn generate() -> Self {
+        static GENERATOR: AtomicU32 = AtomicU32::new(1);
```

Comment:
Do we really need atomics?

Reply:
I think this can be a waste

Reply:
I think the value should be unique even if the scope is different.
This is because the values may overlap.

Reply:
I mean, the analyzer is single threaded anyway, and we don't have to worry about conflicts between two or more modules

Reply:
Come to think of it, I think it would be good to go back to the parent scope from the current scope to find the parent.

Postscript
There was a case of concern, but I can't remember.

Reply:
We can use a monotonicallly increasing numbe by using `Rc<Cell<u32>>`

Reply:
Does your suggestion use the number of pointers in Rc as a counter?

Reply:
No, `Cell<u32>` is the value

Reply:
Is this what you wanted?

```rs
impl DestructureId {
    thread_local! {
        static LAST_ID: Rc<Cell<u32>> = Rc::new(Cell::new(0));
    }

    pub fn generate() -> Self {
        let id = DestructureId::LAST_ID.with(|id| {
            let last_id = id.get().add(1);
            id.set(last_id);
            last_id
        });

        DestructureId(id)
    }

    pub fn get(id: u32) -> Self {
        DestructureId(id)
    }

    pub fn extract(&self) -> u32 {
        self.0
    }
}
```

Reply:
I meant storing a counter in the `Analyzer`

Reply:
I will proceed after finishing the tuple.
thx

Reply:
fixed it